share/sql/public_link.go

```diff
@@ -0,0 +1,417 @@
+	case link.UpdatePublicShareRequest_Update_TYPE_EXPIRATION:
+		res = m.db.Model(&publiclink).Update("expiration", time.Unix(int64(req.Update.GetGrant().Expiration.Seconds), 0))
+	case link.UpdatePublicShareRequest_Update_TYPE_PASSWORD:
+		if req.Update.GetGrant().Password == "" {
+			// Remove the password
+			res = m.db.Model(&publiclink).Update("password", "")
+		} else {
+			// Update the password
+			hashedPwd, err := hashPassword(req.Update.GetGrant().Password, m.c.LinkPasswordHashCost)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not hash share password")
+			}
+			res = m.db.Model(&publiclink).Update("password", hashedPwd)
+		}
+	case link.UpdatePublicShareRequest_Update_TYPE_DESCRIPTION:
+		res = m.db.Model(&publiclink).Update("description", req.Update.GetDescription())
+	case link.UpdatePublicShareRequest_Update_TYPE_NOTIFYUPLOADS:
+		res = m.db.Model(&publiclink).Update("notify_uploads", req.Update.GetNotifyUploads())
+	case link.UpdatePublicShareRequest_Update_TYPE_NOTIFYUPLOADSEXTRARECIPIENTS:
+		res = m.db.Model(&publiclink).Update("notify_uploads_extra_recipients", req.Update.GetNotifyUploadsExtraRecipients())
+	default:
+		return nil, fmt.Errorf("invalid update type: %v", req.GetUpdate().GetType())
+	}
+	if res.Error != nil {
+		return nil, res.Error
+	}
+
+	return m.GetPublicShare(ctx, u, req.Ref, true)
+
+}
+
+func (m *publicShareMgr) GetPublicShare(ctx context.Context, u *user.User, ref *link.PublicShareReference, sign bool) (*link.PublicShare, error) {
+	var ln *model.PublicLink
+	var err error
+	switch {
+	case ref.GetId() != nil:
+		ln, err = m.getLinkByID(ctx, ref.GetId())
+	case ref.GetToken() != "":
+		ln, err = m.getLinkByToken(ctx, ref.GetToken())
+	default:
+		err = errtypes.NotFound(ref.String())
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	l := ln.AsCS3PublicShare()
+	if ln.Password != "" && sign {
+		if err := publicshare.AddSignature(l, ln.Password); err != nil {
+			return nil, err
+		}
+	}
+
+	return l, nil
+}
+
+func (m *publicShareMgr) ListPublicShares(ctx context.Context, u *user.User, filters []*link.ListPublicSharesRequest_Filter, md *provider.ResourceInfo, sign bool) ([]*link.PublicShare, error) {
+	query := m.db.Model(&model.PublicLink{}).
+		Where("orphan = ?", false)
+
+	// TODO: should we enforce that some filters are set here?
+	// Otherwise you can just list all public shares...
+
+	// Append filters
+	m.appendLinkFiltersToQuery(query, filters)
+
+	var links []model.PublicLink
+	var cs3links []*link.PublicShare
+	res := query.Find(&links)
+	if res.Error != nil {
+		return nil, res.Error
+	}
+
+	for _, l := range links {
+		if !isExpired(l) {
+			cs3links = append(cs3links, l.AsCS3PublicShare())
+		}
+	}
+
+	return cs3links, nil
+}
+
+func (m *publicShareMgr) RevokePublicShare(ctx context.Context, u *user.User, ref *link.PublicShareReference) error {
+	var err error
+	var publiclink *model.PublicLink
+	// We need to actually get the link to make sure it is not expired
```

Comment:
In many places you make this comment. Why do we need to make sure is not expired? Do we care?